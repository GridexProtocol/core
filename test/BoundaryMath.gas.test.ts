import {ethers} from "hardhat";
import {MAX_BOUNDARY, MAX_RATIO, MIN_BOUNDARY, MIN_RATIO, Resolution} from "./shared/util";
import {loadFixture} from "@nomicfoundation/hardhat-network-helpers";
import {expect} from "./shared/expect";

describe("BoundaryMath", () => {
    const deployFixture = async function () {
        const boundaryMathFactory = await ethers.getContractFactory("BoundaryMathTest");
        const boundaryMath = await boundaryMathFactory.deploy();
        await boundaryMath.deployed();
        return {boundaryMath};
    };

    describe("#getPriceX96AtBoundary", () => {
        const boundaries = [
            MIN_BOUNDARY,
            -500000,
            -400000,
            -300000,
            -200000,
            -100000,
            0,
            100000,
            200000,
            300000,
            400000,
            MAX_BOUNDARY,
        ];

        for (const boundary of boundaries) {
            it(`boundary = ${boundary}`, async () => {
                const {boundaryMath} = await loadFixture(deployFixture);
                const {priceX96, gasUsed} = await boundaryMath.getPriceX96AtBoundaryWithGasUsed(boundary);
                const cmpObj = {
                    priceX96: priceX96.toBigInt(),
                    gasUsed: gasUsed.toNumber(),
                };
                expect(cmpObj).toMatchSnapshot();
            });
        }
    });

    describe("#getBoundaryAtPriceX96", () => {
        const priceX96s = [
            MIN_RATIO,
            15319379n,
            337263108622n,
            7425001144658883n,
            163464786360687385626n,
            3598751819609688046946419n,
            79228162514264337593543950336n,
            1744244129640337381386292603617838n,
            38400329974042030913961448288742562464n,
            845400776793423922697130608897531771147615n,
            18611883644907511909590774894315720731532604461n,
            MAX_RATIO,
        ];
        for (const priceX96 of priceX96s) {
            it(`priceX96 = ${priceX96}`, async () => {
                const {boundaryMath} = await loadFixture(deployFixture);
                const {boundary, gasUsed} = await boundaryMath.getBoundaryAtPriceX96WithGasUsed(priceX96);
                const cmpObj = {
                    boundary: boundary,
                    gasUsed: gasUsed.toNumber(),
                };
                expect(cmpObj).toMatchSnapshot();
            });
        }
    });

    describe("#getBoundaryLowerAtBoundary", () => {
        const tests = [
            {
                boundary: MIN_BOUNDARY,
                resolution: Resolution.LOW,
            },
            {
                boundary: 0,
                resolution: Resolution.LOW,
            },
            {
                boundary: MAX_BOUNDARY,
                resolution: Resolution.LOW,
            },

            {
                boundary: 3,
                resolution: Resolution.LOW,
            },
            {
                boundary: 3,
                resolution: Resolution.MEDIUM,
            },
            {
                boundary: 3,
                resolution: Resolution.HIGH,
            },
        ];

        for (const test of tests) {
            it(`boundary = ${test.boundary}, resolution = ${test.resolution}`, async () => {
                const {boundaryMath} = await loadFixture(deployFixture);
                const {boundaryLower, gasUsed} = await boundaryMath.getBoundaryLowerAtBoundaryWithGasUsed(
                    test.boundary,
                    test.resolution
                );
                const cmpObj = {
                    boundaryLower: boundaryLower,
                    gasUsed: gasUsed.toNumber(),
                };
                expect(cmpObj).toMatchSnapshot();
            });
        }
    });
});
