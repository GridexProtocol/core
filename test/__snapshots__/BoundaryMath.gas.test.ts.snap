// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoundaryMath #getBoundaryAtPriceX96 priceX96 = 989314 1`] = `
Object {
  "boundary": -527400,
  "gasUsed": 2098,
}
`;

exports[`BoundaryMath #getBoundaryAtPriceX96 priceX96 = 15319379 1`] = `
Object {
  "boundary": -500000,
  "gasUsed": 2126,
}
`;

exports[`BoundaryMath #getBoundaryAtPriceX96 priceX96 = 337263108622 1`] = `
Object {
  "boundary": -400000,
  "gasUsed": 2112,
}
`;

exports[`BoundaryMath #getBoundaryAtPriceX96 priceX96 = 7425001144658883 1`] = `
Object {
  "boundary": -300000,
  "gasUsed": 2140,
}
`;

exports[`BoundaryMath #getBoundaryAtPriceX96 priceX96 = 163464786360687385626 1`] = `
Object {
  "boundary": -200000,
  "gasUsed": 2112,
}
`;

exports[`BoundaryMath #getBoundaryAtPriceX96 priceX96 = 3598751819609688046946419 1`] = `
Object {
  "boundary": -100000,
  "gasUsed": 2112,
}
`;

exports[`BoundaryMath #getBoundaryAtPriceX96 priceX96 = 79228162514264337593543950336 1`] = `
Object {
  "boundary": 0,
  "gasUsed": 2040,
}
`;

exports[`BoundaryMath #getBoundaryAtPriceX96 priceX96 = 1744244129640337381386292603617838 1`] = `
Object {
  "boundary": 100000,
  "gasUsed": 2157,
}
`;

exports[`BoundaryMath #getBoundaryAtPriceX96 priceX96 = 38400329974042030913961448288742562464 1`] = `
Object {
  "boundary": 200000,
  "gasUsed": 2157,
}
`;

exports[`BoundaryMath #getBoundaryAtPriceX96 priceX96 = 845400776793423922697130608897531771147615 1`] = `
Object {
  "boundary": 300000,
  "gasUsed": 2185,
}
`;

exports[`BoundaryMath #getBoundaryAtPriceX96 priceX96 = 18611883644907511909590774894315720731532604461 1`] = `
Object {
  "boundary": 400000,
  "gasUsed": 2157,
}
`;

exports[`BoundaryMath #getBoundaryAtPriceX96 priceX96 = 1461300573427867316570072651998408279850435624081 1`] = `
Object {
  "boundary": 443635,
  "gasUsed": 2203,
}
`;

exports[`BoundaryMath #getBoundaryLowerAtBoundary boundary = -527400, resolution = 1 1`] = `
Object {
  "boundaryLower": -527400,
  "gasUsed": 155,
}
`;

exports[`BoundaryMath #getBoundaryLowerAtBoundary boundary = 0, resolution = 1 1`] = `
Object {
  "boundaryLower": 0,
  "gasUsed": 155,
}
`;

exports[`BoundaryMath #getBoundaryLowerAtBoundary boundary = 3, resolution = 1 1`] = `
Object {
  "boundaryLower": 3,
  "gasUsed": 155,
}
`;

exports[`BoundaryMath #getBoundaryLowerAtBoundary boundary = 3, resolution = 5 1`] = `
Object {
  "boundaryLower": 0,
  "gasUsed": 155,
}
`;

exports[`BoundaryMath #getBoundaryLowerAtBoundary boundary = 3, resolution = 30 1`] = `
Object {
  "boundaryLower": 0,
  "gasUsed": 155,
}
`;

exports[`BoundaryMath #getBoundaryLowerAtBoundary boundary = 443635, resolution = 1 1`] = `
Object {
  "boundaryLower": 443635,
  "gasUsed": 155,
}
`;

exports[`BoundaryMath #getPriceX96AtBoundary boundary = -100000 1`] = `
Object {
  "gasUsed": 792,
  "priceX96": 3598751819609688046946419n,
}
`;

exports[`BoundaryMath #getPriceX96AtBoundary boundary = -200000 1`] = `
Object {
  "gasUsed": 792,
  "priceX96": 163464786360687385626n,
}
`;

exports[`BoundaryMath #getPriceX96AtBoundary boundary = -300000 1`] = `
Object {
  "gasUsed": 820,
  "priceX96": 7425001144658883n,
}
`;

exports[`BoundaryMath #getPriceX96AtBoundary boundary = -400000 1`] = `
Object {
  "gasUsed": 792,
  "priceX96": 337263108622n,
}
`;

exports[`BoundaryMath #getPriceX96AtBoundary boundary = -500000 1`] = `
Object {
  "gasUsed": 806,
  "priceX96": 15319379n,
}
`;

exports[`BoundaryMath #getPriceX96AtBoundary boundary = -527400 1`] = `
Object {
  "gasUsed": 778,
  "priceX96": 989314n,
}
`;

exports[`BoundaryMath #getPriceX96AtBoundary boundary = 0 1`] = `
Object {
  "gasUsed": 710,
  "priceX96": 79228162514264337593543950336n,
}
`;

exports[`BoundaryMath #getPriceX96AtBoundary boundary = 100000 1`] = `
Object {
  "gasUsed": 827,
  "priceX96": 1744244129640337381386292603617838n,
}
`;

exports[`BoundaryMath #getPriceX96AtBoundary boundary = 200000 1`] = `
Object {
  "gasUsed": 827,
  "priceX96": 38400329974042030913961448288742562464n,
}
`;

exports[`BoundaryMath #getPriceX96AtBoundary boundary = 300000 1`] = `
Object {
  "gasUsed": 855,
  "priceX96": 845400776793423922697130608897531771147615n,
}
`;

exports[`BoundaryMath #getPriceX96AtBoundary boundary = 400000 1`] = `
Object {
  "gasUsed": 827,
  "priceX96": 18611883644907511909590774894315720731532604461n,
}
`;

exports[`BoundaryMath #getPriceX96AtBoundary boundary = 443635 1`] = `
Object {
  "gasUsed": 873,
  "priceX96": 1461300573427867316570072651998408279850435624081n,
}
`;
